rclrs/src/wait/guard_condition.rs

```diff
@@ -0,0 +1,156 @@
+use crate::rcl_bindings::*;
+use crate::{Context, RclrsError, ToResult};
+
+use std::sync::{atomic::AtomicBool, Arc, Mutex};
+
+/// A waitiable entity used for waking up a wait set manually
+///
+/// # Example
+/// ```
+/// # use rclrs::{Context, GuardCondition, WaitSet, RclrsError};
+/// # use std::sync::{Arc, atomic::Ordering};
+///
+/// let context = Context::new([])?;
+///
+/// let atomic_bool = Arc::new(std::sync::atomic::AtomicBool::new(false));
+/// let atomic_bool_for_closure = Arc::clone(&atomic_bool);
+///
+/// let gc = Arc::new(GuardCondition::new(
+///     &context,
+///     Some(Box::new(move || {
+///         atomic_bool_for_closure.store(true, Ordering::Relaxed);
+///     })),
+/// ));
+///
+/// let mut ws = WaitSet::new(0, 1, 0, 0, 0, 0, &context)?;
+/// ws.add_guard_condition(Arc::clone(&gc))?;
+///
+/// // Trigger the guard condition, firing the callback and waking the wait set being waited on, if any.
+/// gc.trigger()?;
+///
+/// // The provided callback has now been called.
+/// assert_eq!(atomic_bool.load(Ordering::Relaxed), true);
+///
+/// // The wait call will now immediately return.
+/// ws.wait(Some(std::time::Duration::from_millis(10)))?;
+///
+/// # Ok::<(), RclrsError>(())
+/// ```
+pub struct GuardCondition {
+    /// The rcl_guard_condition_t that this struct encapsulates.
+    pub(crate) rcl_guard_condition: Arc<Mutex<rcl_guard_condition_t>>,
+    /// An optional callback to call when this guard condition is triggered.
+    callback: Option<Box<dyn Fn() + Send + Sync>>,
+    /// A flag to indicate if this guard condition has already been assigned to a wait set.
+    pub(crate) in_use_by_wait_set: Arc<AtomicBool>,
+}
+
+impl Drop for GuardCondition {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: No precondition for this function (besides passing in a valid guard condition)
+            rcl_guard_condition_fini(&mut *self.rcl_guard_condition.lock().unwrap());
+        }
+    }
+}
+
+// SAFETY: rcl_guard_condition is the only member that doesn't implement Send, and it is designed to be accessed from other threads
+unsafe impl Send for rcl_guard_condition_t {}
+
+// SAFETY: The mutexes and atomic members ensure synchronized access to members, and the callback is reentrant
+unsafe impl Sync for GuardCondition {}
+
+impl GuardCondition {
+    /// Creates a new guard condition.
+    pub fn new(context: &Context, callback: Option<Box<dyn Fn() + Send + Sync>>) -> Arc<Self> {
```

Comment:
I disagree about returning an `Arc` favoring usability. For me, conventions and the principle of least surprise make APIs more usable, in this case, given that a `GuardCondition` behaves like other waitable entities, I'd return a `Self` here and add a `create_guard_condition` method to `Node`, which would return an `Arc`.

In this case `rclcpp` is being inconsistent, and `rclpy` fixed this inconsistency https://github.com/ros2/rclpy/blob/master/rclpy/rclpy/node.py#L1696

Reply:
All the other user-facing methods for creating waitable entities (`Node::create_subscription` etc.) already return `Arc`, so if you want `GuardCondition` to behave like other waitable entities, that's actually in favor of making this method return an `Arc` too.

Python is not really comparable, everything is an `Arc` (or at least `Rc`) there basically.

Reply:
> All the other user-facing methods for creating waitable entities (`Node::create_subscription` etc.) already return `Arc`, so if you want `GuardCondition` to behave like other waitable entities, that's actually in favor of making this method return an `Arc` too.

No, because `Subscription::new` for example, returns a `Self`, not an `Arc<Self>`. `create_X` are not comparable with `::new` as they need to add the waitable to an internal list in the `Node`

> Python is not really comparable, everything is an `Arc` (or at least `Rc`) there basically.

I was comparing it at the API level, not at the functional level, I thought that was implied.

Reply:
> > All the other user-facing methods for creating waitable entities (`Node::create_subscription` etc.) already return `Arc`, so if you want `GuardCondition` to behave like other waitable entities, that's actually in favor of making this method return an `Arc` too.
> 
> No, because `Subscription::new` for example, returns a `Self`, not an `Arc<Self>`. `create_X` are not comparable with `::new` as they need to add the waitable to an internal list in the `Node`

That's what I meant with "user-facing" – `Subscription::new` etc. are private.

But this touches on an interesting point: I would say `GuardCondition::new` can be public, since a guard condition doesn't always need to be attached to a node, does it? There _might_ be a use case of attaching it to a node, if you don't `spin()` the node but add it to a wait set, but there is also a use for using a standalone guard condition with a wait set, i.e. where the guard condition is external to the node in the wait set.

> 
> > Python is not really comparable, everything is an `Arc` (or at least `Rc`) there basically.
> 
> I was comparing it at the API level, not at the functional level, I thought that was implied.

Hmm, I'm not sure what you're getting at with the reference to `rclpy`, sorry ... might need to read up on guard conditions in `rclcpp` and `rclpy`.

Reply:
I wouldn't mind your suggestion of a free function, but it would look a little strange. I think @esteve should let us know if he still thinks the current state of things inconsistent after considering the points on private vs public, and the use case for guard conditions that are _not_ part of the node.

Otherwise, the next best thing imo would be to add both the `GuardCondition::new() -> Self` version and a `GuardCondition::new_arc() -> Arc<Self>` version.

Then there is the other, unrelated point of adding `Node::create_guard_condition()`. I agree with Esteve here that we probably want it – sure, guard conditions are not logically part of a node, but it could be potentially useful (and `rclpy` has such a method). E.g. for the dynamic message GUI I had only one node, and I added that node and the guard condition to the wait set separately. With this added method (which would store a `Weak<GuardCondition>` inside the node), the guard condition could be added to the wait set as part of adding the node to the wait set, making the code shorter. We need to clearly document the difference between `Node::create_guard_condition()` and `GuardCondition::new(_arc)()` though.

Of course `Node::create_guard_condition()` should return an `Arc`, I think we all agree there.

Reply:
It's strange to me that there's no `Node::create_guard_condition()` in `rclcpp`, but there is in `rclpy`. In this case I'd favor having that method to make the API more consistent with the rest of waitable entities.

`rclpy` even keeps track of the guard conditions, which are triggered on every `spin`, so this inconsistency between `rclpy` and `rclcpp` is rather intriguing.

Reply:
@christopherjreid, I agree with Esteve, please add `Node::create_guard_condition()`.

Reply:
@esteve you haven't commented on this yet: 

> I think @esteve should let us know if he still thinks the current state of things (i.e. returning an `Arc` from `GuardCondition::new()`) inconsistent after considering the points on private vs public, and the use case for guard conditions that are _not_ part of the node.

Reply:
I'd still prefer `GuardCondition::new` to return a `GuardCondition` to keep it consistent with the other waitables, and then give users `Node::create_guard_condition` (which returns an `Arc`) as the entry point for the API.

Reply:
Ok. thanks, I'll resolve this then. Note though that `Node::create_guard_condition` has different behavior than `Arc::new(GuardCondition::new())` – with the latter, the guard condition is not attached to the node.